tests/functional/utils/convergence_analysis.py

```diff
@@ -0,0 +1,313 @@
+        """Number of levels of the convergence analysis."""
+
+        self.in_space: bool = in_space
+        """Whether a spatial analysis should be performed."""
+
+        self.in_time: bool = in_time
+        """Whether a temporal analysis should be performed."""
+
+        self.spatial_rate: int = spatial_rate
+        """Rate at which the mesh size should be refined. A value of ``2``
+        corresponds to halving the mesh size(s) between :attr:`levels`.
+
+        """
+
+        self.temporal_rate: int = temporal_rate
+        """Rate at which the time step size should be refined. A value of ``2``
+        corresponds to halving the time step size between :attr:`levels`.
+
+        """
+
+        # Initialize setup and retrieve spatial and temporal data
+        setup = model_class(model_params)
+        setup.prepare_simulation()
+
+        self._init_setup = setup
+        """"Initial setup."""
+
+        self._is_time_dependent: bool = setup._is_time_dependent()
+        """Whether the problem is time-dependent or not."""
+
+        # Sanity check
+        if not self._is_time_dependent and self.in_time:
+            raise ValueError("Analysis in time not available for stationary models.")
+
+        # Check whether the grid is simplicial or not
+        g = setup.mdg.subdomains()[0]
+        if isinstance(g, pp.TriangleGrid) or isinstance(g, pp.TetrahedralGrid):
+            self._is_simplicial = True
+        else:
+            self._is_simplicial = False
+
+        # Retrieve list of mesh arguments
+        list_of_meshing_arguments: list[
+            Union[dict[str, pp.number], np.ndarray]
+        ] = self._get_list_of_meshing_arguments()
+
+        # Retrieve list of time managers
+        list_of_time_managers: Union[
+            list[pp.TimeManager], None
+        ] = self._get_list_of_time_managers()
+
+        # Create list of model parameters
+        list_of_params: list[dict] = []
+        for lvl in range(self.levels):
+            params = deepcopy(model_params)
+            params["meshing_arguments"] = list_of_meshing_arguments[lvl]
+            if list_of_time_managers is not None:
+                params["time_manager"] = list_of_time_managers[lvl]
+            list_of_params.append(params)
+
+        self.model_params: list[dict] = list_of_params
+        """List of model parameters associated to each run."""
+
+    def run_analysis(self) -> list:
+        """Run convergence analysis."""
+        convergence_results: list = []
+        for level in range(self.levels):
+            setup = self.model_class(self.model_params[level])
+            if not setup._is_time_dependent():
+                # Run stationary model
+                pp.run_stationary_model(setup, self.model_params[level])
+                # Complement information in results
+                setattr(setup.results[-1], "num_dofs", setup.equation_system.num_dofs())
+                setattr(setup.results[-1], "cell_diam", setup.mdg.diameter())
```

Comment:
I think I'd prefer spelling out diameter. That should make the link to mdg.diameter a bit easier to see.

Reply:
Suggestion incorporated.